### DISCUSSION TITLE
Novice seeking help

### DISCUSSION THREAD

**Discussion by yunfengwangluo** [emilk/egui#3227] [Q&A]
Help!

In EGUI, I have another thread that needs to continuously update a "vec! [..]" data, which needs to be synchronously updated and displayed in the EGUI window. Do you have any good ideas or similar examples for reference.

The 'unsafe' method can also be accepted without affecting performance.

Feeling endless!
eframe="0.22.0"

**Comment by mattfbacon** [comment]
I suggest asking this on the Rust users forum at https://users.rust-lang.org

**Comment by podusowski** [comment]
Just share it with `Arc<Mutex/RwLock>>` along with `egui::Context`. When you change it, call `request_repaint` on the context and in GUI thread just from it.

**Comment by yunfengwangluo** [comment]
> Just share it with `Arc<Mutex/RwLock>>` along with `egui::Context`. When you change it, call `request_repaint` on the context and in GUI thread just from it.

I am also considering using 'Arc<Mutex/RwLock>>'. But we always encounter work issues, such as ownership and lifecycle issues.

If it's convenient for you, could you provide an example of home! Let me refer to it. Thank you very much!!!

**Comment by gitsmol** [comment]
Another way is to use a mpsc channel:
Main thread (ui):
- try_recv in main loop (mind that you force updates occasionally with request_repaint_after_ms, otherwise this only runs when you interact with egui):
```if let Ok(data) = rx.try_recv() { update the struct with the data egui displays }```

worker thread:
- do data processing stuff
- when ready, send new data:
```tx.send(mydata)```

[The Rust book provides a useful introduction of this concept.](https://doc.rust-lang.org/book/ch16-02-message-passing.html)

You can easily create the tx/rx pair on egui init and then pass off a clone of the tx to the worker thread. I find this solution to be a little more friendly as it doesn't rely on locking the mutex and (in some implementations) freezing either the main thread or the helper thread. Because the problem with the mutex is, in my experience, that in order to check if you have any updates inside that mutex, you have to lock it for the other thread. When egui checks that mutex 60 times per second (60 fps), the other thread has to wait very often.

**Comment by mattfbacon** [comment]
> mind that you force updates occasionally with request_repaint_after_ms, otherwise this only runs when you interact with egui

Much better to call `Context::request_repaint` in the sender to avoid unnecessary repaints and latency. `Context` is `Send` after all.

**Comment by gitsmol** [comment]
> > mind that you force updates occasionally with request_repaint_after_ms, otherwise this only runs when you interact with egui
> 
> Much better to call `Context::request_repaint` in the sender to avoid unnecessary repaints and latency. `Context` is `Send` after all.

Yes, you are right. Learned something :)

**Comment by yunfengwangluo** [comment]
> Another way is to use a mpsc channel: Main thread (ui):
> 
> * try_recv in main loop (mind that you force updates occasionally with request_repaint_after_ms, otherwise this only runs when you interact with egui):
>   `if let Ok(data) = rx.try_recv() { update the struct with the data egui displays }`
> 
> worker thread:
> 
> * do data processing stuff
> * when ready, send new data:
>   `tx.send(mydata)`
> 
> [The Rust book provides a useful introduction of this concept.](https://doc.rust-lang.org/book/ch16-02-message-passing.html)
> 
> You can easily create the tx/rx pair on egui init and then pass off a clone of the tx to the worker thread. I find this solution to be a little more friendly as it doesn't rely on locking the mutex and (in some implementations) freezing either the main thread or the helper thread. Because the problem with the mutex is, in my experience, that in order to check if you have any updates inside that mutex, you have to lock it for the other thread. When egui checks that mutex 60 times per second (60 fps), the other thread has to wait very often.

Thank you for your help, your method is very effective! My basic skills are too poor, I still need to continue learning!!!

Thank you!
